src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+  private void handleMalformedDocResponse(BulkItemResponse response) {
+    String errorMsg = String.format(
+        "Encountered an illegal document error '%s'. Ignoring and will not index record.",
+        response.getFailureMessage()
+    );
+    switch (config.behaviorOnMalformedDoc()) {
+      case IGNORE:
+        log.debug(errorMsg);
+        return;
+      case WARN:
+        log.warn(errorMsg);
+        return;
+      case FAIL:
+      default:
+        log.error(
+            "Encountered an illegal document error '{}'. To ignore future records like this,"
+                + " change the configuration '{}' to '{}'.",
+            response.getFailureMessage(),
+            ElasticsearchSinkConnectorConfig.BEHAVIOR_ON_MALFORMED_DOCS_CONFIG,
+            BehaviorOnMalformedDoc.IGNORE
+        );
+        error.compareAndSet(
+            null,
+            new ConnectException("Indexing record failed.", response.getFailure().getCause())
+        );
+    }
   }
 
   /**
-   * Perform a refresh of all indexes, making all indexed data searchable (useful mostly for test)
+   * Whether there is a failed response.
    *
-   * @throws IOException If the client cannot execute the request
+   * @return true if a response has failed, false if none have failed
    */
-  default void refresh() throws IOException {
-    throw new UnsupportedOperationException("refresh is not implemented yet by this client");
+  private boolean isFailed() {
+    return error.get() != null;
   }
 
   /**
-   * Shuts down the client.
+   * Removes the mapping for document id to record being written.
+   *
+   * @param id the document id
    */
-  void close();
+  private void removeFromRecordMap(String id) {
+    if (recordMap != null) {
+      recordMap.remove(id);
+    }
+  }
+
+  /**
+   * Reports a bad record to the DLQ.
+   *
+   * @param response the failed response from ES
+   */
+  private synchronized void reportBadRecord(BulkItemResponse response) {
+    if (reporter != null) {
+      SinkRecord original = recordMap.get(response.getId());
+      if (original != null) {
+        reporter.report(
+            original,
+            new ReportingException("Indexing failed: " + response.getFailureMessage())
+        );
+      }
+    }
+  }
+
+  /**
+   * Exception that swallows the stack trace used for reporting errors from Elasticsearch
+   * (mapper_parser_exception, illegal_argument_exception, and action_request_validation_exception)
+   * resulting from bad records using the AK 2.6 reporter DLQ interface.
+   */
+  @SuppressWarnings("serial")
+  public static class ReportingException extends RuntimeException {
+
+    public ReportingException(String message) {
+      super(message);
+    }
+
+    /**
+     * This method is overriden to swallow the stack trace.
+     *
+     * @return Throwable
+     */
+    @Override
+    public synchronized Throwable fillInStackTrace() {
```

Comment:
why is this `synchronized` ?

Reply:
the method it is overriding is `synchronized`

Reply:
You don't mutate anything. Fine to keep the contract though.